File: sdk/typescript/rooch-sdk/src/bcs/args.ts
// Copyright (c) RoochNetwork
// SPDX-License-Identifier: Apache-2.0

import { bcs, SerializedBcs } from '@mysten/bcs'

import { Bytes } from '../types/bytes.js'
import { isBytes, toHEX } from '../utils/index.js'
import { u16, u32, u8, u64, u128, u256, bool, address, objectId } from '../types/index.js'

import { Address, ObjectId } from './bcs.js'
import { Serializer } from './serializer.js'
import { StructTag } from './types.js'

export type ArgType =
  | 'u8'
  | 'u16'
  | 'u32'
  | 'u64'
  | 'u128'
  | 'u256'
  | 'bool'
  | 'string'
  | 'object'
  | 'objectId'
  | 'address'

export class Args {
  readonly value: Bytes

  constructor(input: Bytes) {
    this.value = input
  }

  encodeWithHex(): string {
    return toHEX(this.value)
  }

  encode(): Bytes {
    return this.value
  }

  static u8(input: u8) {
    return new Args(bcs.u8().serialize(input).toBytes())
  }

  static u16(input: u16) {
    return new Args(bcs.u16().serialize(input).toBytes())
  }

  static u32(input: u32) {
    return new Args(bcs.u32().serialize(input).toBytes())
  }

  static u64(input: u64) {
    return new Args(bcs.u64().serialize(input).toBytes())
  }

  static u128(input: u128) {
    return new Args(bcs.u128().serialize(input).toBytes())
  }

  static u256(input: u256) {
    return new Args(bcs.u256().serialize(input).toBytes())
  }

  static bool(input: bool) {
    return new Args(bcs.bool().serialize(input).toBytes())
  }

  static string(input: string) {
    return new Args(bcs.string().serialize(input).toBytes())
  }

  static address(input: address) {
    return new Args(Address.serialize(input).toBytes())
  }

  static object(input: StructTag) {
    return this.objectId(Serializer.structTagToObjectID(input))
  }

  static objectId(input: objectId) {
    return new Args(ObjectId.serialize(input).toBytes())
  }

  static struct(input: SerializedBcs<any> | Bytes): Args {
    return new Args(isBytes(input) ? input : input.toBytes())
  }

  static vec(type: ArgType, input: number[] | bigint[] | boolean[] | string[] | StructTag[]) {
    let _value: Bytes
    switch (type) {
      case 'u8':
        _value = bcs
          .vector(bcs.u8())
          .serialize(input as number[])
          .toBytes()
        break
      case 'u16':
        _value = bcs
          .vector(bcs.u16())
          .serialize(input as number[])
          .toBytes()
        break
      case 'u32':
        _value = bcs
          .vector(bcs.u32())
          .serialize(input as number[])
          .toBytes()
        break
      case 'u64':
        _value = bcs
          .vector(bcs.u64())
          .serialize(input as bigint[])
          .toBytes()
        break
      case 'u128':
        _value = bcs
          .vector(bcs.u128())
          .serialize(input as bigint[])
          .toBytes()
        break
      case 'u256':
        _value = bcs
          .vector(bcs.u256())
          .serialize(input as number[])
          .toBytes()
        break
      case 'bool':
        _value = bcs
          .vector(bcs.bool())
          .serialize(input as boolean[])
          .toBytes()
        break
      case 'string':
        _value = bcs
          .vector(bcs.string())
          .serialize(input as string[])
          .toBytes()
        break
      case 'object':
        const tmp = (input as StructTag[]).map(Serializer.structTagToObjectID)
        _value = bcs.vector(ObjectId).serialize(tmp).toBytes()
        break
      case 'objectId':
        _value = bcs
          .vector(ObjectId)
          .serialize(input as string[])
          .toBytes()
        break
      case 'address':
        _value = bcs
          .vector(Address)
          .serialize(input as string[])
          .toBytes()
        break
    }

    return new Args(_value)
  }
}

File: sdk/golang/rooch-sdk/bcs/args.go
package bcs

import (
	"encoding/hex"

	"github.com/rooch-network/rooch/sdk/golang/rooch-sdk/types"
	"github.com/rooch-network/rooch/sdk/golang/rooch-sdk/utils"
)

type ArgType string

const (
	ArgTypeU8       ArgType = "u8"
	ArgTypeU16      ArgType = "u16"
	ArgTypeU32      ArgType = "u32"
	ArgTypeU64      ArgType = "u64"
	ArgTypeU128     ArgType = "u128"
	ArgTypeU256     ArgType = "u256"
	ArgTypeBool     ArgType = "bool"
	ArgTypeString   ArgType = "string"
	ArgTypeObject   ArgType = "object"
	ArgTypeObjectId ArgType = "objectId"
	ArgTypeAddress  ArgType = "address"
)

type Args struct {
	value []byte
}

func NewArgs(input []byte) *Args {
	return &Args{value: input}
}

func (a *Args) EncodeWithHex() string {
	return hex.EncodeToString(a.value)
}

func (a *Args) Encode() []byte {
	return a.value
}

func U8(input uint8) *Args {
	return NewArgs(utils.SerializeU8(input))
}

func U16(input uint16) *Args {
	return NewArgs(utils.SerializeU16(input))
}

func U32(input uint32) *Args {
	return NewArgs(utils.SerializeU32(input))
}

func U64(input uint64) *Args {
	return NewArgs(utils.SerializeU64(input))
}

func U128(input *types.U128) *Args {
	return NewArgs(utils.SerializeU128(input))
}

func U256(input *types.U256) *Args {
	return NewArgs(utils.SerializeU256(input))
}

func Bool(input bool) *Args {
	return NewArgs(utils.SerializeBool(input))
}

func String(input string) *Args {
	return NewArgs(utils.SerializeString(input))
}

func Address(input string) *Args {
	return NewArgs(utils.SerializeAddress(input))
}

func Object(input *StructTag) *Args {
	objectID := StructTagToObjectID(input)
	return ObjectId(objectID)
}

func ObjectId(input string) *Args {
	return NewArgs(utils.SerializeObjectID(input))
}

func Struct(input []byte) *Args {
	return NewArgs(input)
}

func Vec(argType ArgType, input interface{}) *Args {
	var serialized []byte

	switch argType {
	case ArgTypeU8:
		serialized = utils.SerializeVecU8(input.([]uint8))
	case ArgTypeU16:
		serialized = utils.SerializeVecU16(input.([]uint16))
	case ArgTypeU32:
		serialized = utils.SerializeVecU32(input.([]uint32))
	case ArgTypeU64:
		serialized = utils.SerializeVecU64(input.([]uint64))
	case ArgTypeU128:
		serialized = utils.SerializeVecU128(input.([]*types.U128))
	case ArgTypeU256:
		serialized = utils.SerializeVecU256(input.([]*types.U256))
	case ArgTypeBool:
		serialized = utils.SerializeVecBool(input.([]bool))
	case ArgTypeString:
		serialized = utils.SerializeVecString(input.([]string))
	case ArgTypeObject:
		structTags := input.([]*StructTag)
		objectIDs := make([]string, len(structTags))
		for i, tag := range structTags {
			objectIDs[i] = StructTagToObjectID(tag)
		}
		serialized = utils.SerializeVecObjectID(objectIDs)
	case ArgTypeObjectId:
		serialized = utils.SerializeVecObjectID(input.([]string))
	case ArgTypeAddress:
		serialized = utils.SerializeVecAddress(input.([]string))
	}

	return NewArgs(serialized)
} 